src/main/java/org/venice/beachfront/bfapi/auth/OpenIDTokenProvider.java

```diff
@@ -0,0 +1,145 @@
+
+    public OpenIDTokenProvider() {
+        super.setStateKeyGenerator(stateKeyGenerator);
+    }
+
+    @Override
+    public OAuth2AccessToken obtainAccessToken(OAuth2ProtectedResourceDetails details, AccessTokenRequest request)
+        throws UserRedirectRequiredException, UserApprovalRequiredException, AccessDeniedException,
+        OAuth2AccessDeniedException {
+
+        AuthorizationCodeResourceDetails resource = (AuthorizationCodeResourceDetails) details;
+
+        if (request.getAuthorizationCode() == null) {
+            if (request.getStateKey() == null) {
+                throw getRedirectForAuthorization(resource, request);
+            }
+            obtainAuthorizationCode(resource, request);
+        }
+        return retrieveToken(request, resource, getParametersForTokenRequest(resource, request),
+            getHeadersForTokenRequest(request));
+
+    }
+
+    private HttpHeaders getHeadersForTokenRequest(AccessTokenRequest request) {
+        HttpHeaders headers = new HttpHeaders();
+        // No cookie for token request
+        return headers;
+    }
+
+    private MultiValueMap<String, String> getParametersForTokenRequest(AuthorizationCodeResourceDetails resource,
+                                                                       AccessTokenRequest request) {
+
+        MultiValueMap<String, String> form = new LinkedMultiValueMap<String, String>();
+        form.set("grant_type", "authorization_code");
+        form.set("code", request.getAuthorizationCode());
+        form.set("response_type", "id_token");
+
+        Object preservedState = request.getPreservedState();
+        if (request.getStateKey() != null || stateMandatory) {
+            // The token endpoint has no use for the state so we don't send it back, but we are using it
+            // for CSRF detection client side...
+            if (preservedState == null) {
+                throw new InvalidRequestException(
+                    "Possible CSRF detected - state parameter was required but no state could be found");
+            }
+        }
+
+        // Extracting the redirect URI from a saved request should ignore the current URI, so it's not simply a call to
+        // resource.getRedirectUri()
+        String redirectUri = null;
+        // Get the redirect uri from the stored state
+        if (preservedState instanceof String) {
+            // Use the preserved state in preference if it is there
+            // TODO: treat redirect URI as a special kind of state (this is a historical mini hack)
+            redirectUri = String.valueOf(preservedState);
+        }
+        else {
+            redirectUri = resource.getRedirectUri(request);
+        }
+
+        if (redirectUri != null && !"NONE".equals(redirectUri)) {
+            form.set("redirect_uri", redirectUri);
+        }
+
+        return form;
+
+    }
+
+    private UserRedirectRequiredException getRedirectForAuthorization(AuthorizationCodeResourceDetails resource,
```

Comment:
This makes it seem like this is designed to return an error.  Is that right?

Reply:
That is correct. Keep in mind this class is authored by PCF.